src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using CsvHelper;
+using Raven.Client;
+using Raven.Client.Documents.Smuggler;
+using Raven.Client.Util;
+using Raven.Server.Documents;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Smuggler.Documents.Data;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Smuggler.Documents
+{
+    public class CsvStreamSource : ISmugglerSource, IDisposable
+    {
+        private Stream _stream;
+        private DocumentsOperationContext _context;
+        private DatabaseItemType _currentType;
+        private string _collection;
+        private StreamReader _reader;
+        private CsvReader _csvReader;
+        private bool _hasId;
+        private int _idIndex;
+        private bool _hasCollection;
+        private int _collectionIndex;
+        /// <summary>
+        /// This dictionary maps the index of a property to its nested segments.
+        /// </summary>
+        private Dictionary<int, string[]> _nestedPropertyDictionary = null;
+
+        public CsvStreamSource(Stream stream, DocumentsOperationContext context, string collection)
+        {
+            _stream = stream;
+            _context = context;
+            _currentType = DatabaseItemType.Documents;
+            _collection = collection;
+        }      
+
+        public IDisposable Initialize(DatabaseSmugglerOptions options, SmugglerResult result, out long buildVersion)
+        {
+            buildVersion = 40;
+            _reader = new StreamReader(_stream);
+            _csvReader = new CsvReader(_reader);
+            _csvReader.Configuration.Delimiter = ",";            
+            return new DisposableAction(() =>
+            {
+                _reader.Dispose();
+                _csvReader.Dispose();
+            });
+        }
+
+        private bool _headersProcessed;
+
+        private void ProcessFieldsIfNeeded()
+        {
+            if(_headersProcessed)
+                return;            
+            for (var i = 0; i < _csvReader.FieldHeaders.Length; i++)
+            {
+                if (_csvReader.FieldHeaders[i].Equals(Constants.Documents.Metadata.Id))
+                {
+                    _hasId = true;
+                    _idIndex = i;
+                }
+                if (_csvReader.FieldHeaders[i].Equals(Constants.Documents.Metadata.Collection))
+                {
+                    _hasCollection = true;
+                    _collectionIndex = i;
+                }
+                if (_csvReader.FieldHeaders[i][0] == '@')
```

Comment:
Can we have empty field names?

Reply:
we can have an empty value but not an empty header